### ISSUE TITLE
[Android] Is it possible to use system-provided funts?

### ISSUE THREAD

**Issue by andronoob** [kivy/kivy#6106]
I found that [Electrum](https://github.com/spesmilo/electrum) has an unfixed  CJK issue for several years. As long as the language is switched to Chinese/Japanese etc, the characters just turn blank. It's quite annoying. I've done some googling, but I found no answer other than "pack your own font into your app" - but why should every app pack their own font? It's so weird. What if the owner of smartphone customize the font? Should kivy apps "resist"?

I wonder if it's possible to let kivy simply use the system-provided font, so that the long-lasting CJK issue could be resolved?

**Comment by JonasT** [comment]
Others can help better since I barely use kivy, but regarding the general font situation:

I recently researched this out of curiosity and as far as I'm aware, Android can only change DPI for all apps but has no way of setting the font used in all apps. Consequently for what it's worth, when embedding & "hardcoding" the font to a certain choice, I don't think there's any loss of flexibility in user functionality. So at least from that angle it's not a problematic thing to do

**Comment by andronoob** [comment]
@JonasT 
As far as I know, there are a lot of Android vendors who provide functionality allowing the user to customize system-wide font. But, I agree that font customization is a minor issue.
Google once released a set of fonts, a.k.a Noto, which was intended to cover as many locales as it can. It's almost impossible to pack all these fonts into a single APK for merely localization purpose.

**Comment by tito** [comment]
In contrary, are you 100% sure that the font you asked will be there on all system?
On desktop you can reference system font, we could do the same on Android. But if the font you asked is not there, then that's the same, you'll end up with app that are broken.
We don't have any kind of font management/fallback, and per glyph fallback.
Currently there is the Pango text backend that starts to support it, but that's it :/

**Comment by andronoob** [comment]
@tito 
I've came across a lot of Android apps which seemed to be using machine translation. The translated text might not be accurate, but I've never came across "tofu" issues like Kivy.
Also, although I have little knowledge about Android system, I once tried to build my own toy app - I didn't remember that I must specify a font explicitly.
I think the solution to the localization problem is quite simple: use the default font provided by the system - is this impossible for Kivy?

**Comment by andronoob** [comment]
Why should the app specify a font explicitly? Why should an app use complicated "per glyph fallback" at all?
For an American user, if he installed a Japanese app, he might came across the "tofu" issue - but in my opinion it's actually a non-issue. What actually matters is that the real Japanese user only sees "tofus" on the screen.

**Comment by andronoob** [comment]
>On desktop you can reference system font, we could do the same on Android. But if the font you asked is not there, then that's the same, you'll end up with app that are broken.

Sorry, it's **already broken, now, for any CJK user.**

**Comment by JonasT** [comment]
@andronoob yes and just switching mindlessly over to a system provided font may break it for everyone else, so calm down a little.

See, it's not currently how it is because we hate CJK users, but because `SDL_TTF` and other font renderers have quite simple font handling and as @tito already said, there is no fallback handling (so basically right now you can pick one font and if it's not there it won't work, if I got right how kivy works).

>  I think the solution to the localization problem is quite simple: use the default font provided by the system - is this impossible for Kivy?

No, but it would most likely require new fallback handling so it doesn't all just break if there is somehow a problem with fetching a default TTF. Someone needs to implement that and make sure it works properly, it's most likely more than a one-line change.

**Comment by JonasT** [comment]
A more constructive summary: your suggestion to use system fonts for international support where needed is reasonable, kivy just needs to go about this in a way that doesn't break it for others. And that may be more work than just switching to a specific Android system font if fallback handling is needed (and certainly will need to be tested well)

**Comment by andronoob** [comment]
Sorry for my rudeness. But I'm still confused - did the above comments imply that:

(1) Generally, Android system is responsible to handle all the complicated fallback stuff of text rendering job;

(2) No matter how well Android OS could handle it, perfect or broken, it's impossible for Kivy apps to make use of those features provided by Android system directly, so that Kivy has no choice but to re-implement all of those complicated "per-glyph-fallback" itself?

**Comment by JonasT** [comment]
> It's impossible for Kivy apps to make use of those features provided by Android system directly

Yes that is exactly the problem, because kivy uses SDL2 and/or OpenGL which completely bypasses the entire Android UI toolkit, and it doesn't use any of the Android functions for text widgets a normal app would use (because they aren't cross-platform)

**Comment by tito** [comment]
@andronoob You miss one point that will enlight you: the default font don't have all the glyph available in the world. If you talk about CJK, sometimes it's additionnal font that provide the set of characters.

Let's you want to render something "Hello 你好".
1. in the system, you may have 2 differents font used for this. The way the renderer works is checking the availability of the Glyph for the first prefered font, if not available, fallback on the second one, etc.
2. in Kivy using SDL2 and freetype, you point on a single font filename/font alias. There is no way to point on multiples font filenames, as freetype itself doesn't support Glyph fallback. So yes if you use a font that doesn't have both charset, you're stuck.
3. Pango backend leverage this as they support Glyph fallback. So you may have a list of available font, and always have the possibility to use the Glyph from another font even if that wasn't the one used in the first place.

Yes, so that, either we need to implement the per-glyph rendering, and/or push the pango backend on all platform more that will solve this (bonus, it also supports emoji rendering)

**Comment by andronoob** [comment]
@tito
>the default font don't have all the glyph available in the world.

I've already pointed out that:

1. Generally, Android system is responsible to handle all the complicated fallback stuff of text rendering job

2. Even if Andoid can't handle this correctly, it's still not really a problem - for an American user, if he installed a Japanese app, he might came across the "tofu" issue - but in my opinion it's actually a non-issue. What actually matters is that the real Japanese user only sees "tofus" on the screen.

Sorry for copy&paste, but, I'm really, really confused why Kivy cannot just let Android to do its own job.

**Comment by andronoob** [comment]
As a user, I assume that:

1. Basic level of localization support: the user could just install the app & start to use it in his native language.

Kivy is already **broken** at this level.

2. Advanced level of localization support: even if the OS cannot handle font/text rendering correctly, the App could still work flawlessly - but in some scenarios like copy&paste it still inevitably has "tofu" issue.

It seems that Kivy is currently pursuing this.

**Comment by andronoob** [comment]
Additionally, packing TTFs into the APK would obviously make it bloat - although packing several TTFs still won't "have all the glyph available in the world."

**Comment by tito** [comment]
You have a misconception about what the system can or cannot do.

> * Generally, Android system is responsible to handle all the complicated fallback stuff of text rendering job

-> I would say yes, for native application. All the OpenGL application must do their own stuff, and mostly any project that implement their own rendering engine (as us.)
Just have a look, take Chromium, walk through the source code and see by yourself: https://github.com/chromium/chromium/tree/master/ui/gfx

And if you're not native, then you want to implement your own "text renderer". Check https://stackoverflow.com/questions/7165097/render-text-in-purely-native-android-app for example, there is no NDK api for that, we cannot just "use Android".

Saying that "Kivy is currently pursuing this" is misleading, and just throw away the work toward it.
We know what's wrong, and if it would be easy to fix it would already be done.

**Comment by JonasT** [comment]
@andronoob maybe just let us conclude that we know what needs to be done, but it's not trivial. What do we gain from discussing this further?

> Sorry for copy&paste, but, I'm really, really confused why Kivy cannot just let Android to do its own job.

Then maybe you should stop here :smile: as @tito (and me) already explained this is simply due to how NDK/OpenGL/SDL2 applications inherently work. Like it or not that's how things are :woman_shrugging: if you're dragging us in a discussion circle here, that won't magically change it or get it implemented.

If you have a pull request, we'd be very interested. Otherwise, this will be done once someone of us gets around to doing it

**Comment by andronoob** [comment]
Sorry for disturbing (again), but, I'm still curious if this Canvas.drawText method is possible to solve text rendering issue: https://stackoverflow.com/a/4336679/5713110

I know I'm very ignorant, but, it looks very likely that "using Android" is possible.

I noticed that, as some comments had pointed out, this method might suffer from low efficiency - however, it doesn't look like a issue for an app like Electrum to me.

**Comment by andronoob** [comment]
I also found that most Chinese Kivy app developers don't see this as a problem - they are happy to pack font files into the APK.

However, this doens't seem to be the same case of Electrum, because Electrum has to support much more languages.

**Comment by BornForFever** [comment]
However, in China, fonts **vary** with ROM packages. That's to say, fonts used in Huawei devices are **different** from those in Xiaomi. If I package my Xiaomi font into the apk and release, and then a Huawei user installs on his device and sees my Xiaomi fonts, won't he feel weird?

**Comment by akshayaurora** [comment]
Kivy does have support for using system provided fonts, just not for using font fallback or reshaping.

https://github.com/kivy/kivy/blob/master/kivy/core/text/__init__.py#L381

You should be able to specify any font present in the dir /system/fonts on android 

On electrum thus you should be able to specify a font specific to CJK at the time of changing the language, and things should work!

**Comment by akshayaurora** [comment]
for ref:
https://github.com/spesmilo/electrum/issues/2032
https://github.com/spesmilo/electrum/issues/5703

**Comment by andronoob** [comment]
@akshayaurora To my understanding, Android knows which font to use, furthermore, it can combine multiple fonts to render text in complicated scenarios like mixed language (aka "per-glyph fallback"), while Kivy (for Android, which doesn't have/use pango library) can't do this at all (Kivy application has to hard-code font related things) - that's the problem. You mentioned DroidSansFallback.ttf, however it may or may not exist.

On Electrum it's not only about CJK, it's also about many other languages like Arabian.

**Comment by kmAyushkm** [comment]
It is possibke to compile pango for android. I am not sure how it happens but i give some sources..
![Screenshot_20210817-122256_Termux](https://user-images.githubusercontent.com/80582425/129677901-b9ee1828-ff5c-46a2-a4dd-94c3fd5a6e37.jpg)
https://github.com/termux/termux-packages/tree/master/packages/pango
I think by waking some changes in the source code it can be compiled.
